blocks/datasets/__init__.py

```diff
@@ -0,0 +1,233 @@
+        dataset are requested.
+
+    """
+    def __init__(self, data, iteration_scheme=None, sources=None):
+        update_instance(self, locals())
+        self.data_state = None
+
+    def __iter__(self):
+        self.data_state = self.data.open(self.data_state)
+        # TODO Allow for multiple child datasets/data streams, so that they
+        # can be combined
+        if isinstance(self.data, DataStream):
+            self.child_iterator = iter(self.data)
+        return DataIterator(self, iter(self.iteration_scheme)
+                            if self.iteration_scheme else None, self.sources)
+
+    @property
+    def sources(self):
+        if getattr(self, '_sources', None) is None:
+            return self.data.sources
+
+    @sources.setter
+    def sources(self, sources):
+        self._sources = sources
+
+    def open(self, state=None):
+        return self.data.open(state)
+
+    def close(self, satte):
+        self.data.close(self.data_state)
+
+    def get_data(self, state=None, request=None, sources=None):
+        """Get data from the dataset.
+
+        Notes
+        -----
+        This is the default implementation which redirects the request for
+        data directly to the dataset (or wrapped data stream). For more
+        complex data streams, one could perform e.g. caching here.
+
+        """
+        return self.data.get_data(state, request, sources)
+
+
+class CachedDataStream(DataStream):
+    """Cache examples when sequentially reading a dataset.
+
+    Given a data stream which reads large chunks of data, this data
+    stream caches these chunks and returns smaller batches from it until
+    exhausted.
+
+    Parameters
+    ----------
+    iteration_scheme : :class:`IterationScheme`
+        Note that this iteration scheme must return either batch sizes
+        (integers) or sequential batches (slices), which must necessarily
+        be smaller than the child data stream i.e. the batches returned
+        must be smaller than the cache size.
+
+    """
+    def __init__(self, data, iteration_scheme, *args, **kwargs):
+        super(CachedDataStream, self).__init__(data, iteration_scheme, *args,
+                                               **kwargs)
+        self.cache = [[] for source in self.sources]
+
+    def get_data(self, state=None, request=None, sources=None):
+        if isinstance(request, six.integer_types):
+            batch_size = request
+        elif isinstance(request, slice):
+            batch_size = request.stop - request.start
+        if batch_size >= len(self.cache[0]):
+            self._cache(state, sources)
+        data = []
+        for i, cache in enumerate(self.cache):
+            data.append(numpy.asarray(cache[:batch_size]))
+            self.cache[i] = cache[batch_size:]
+        return tuple(data)
+
+    def _cache(self, state, sources):
+        for cache, data in zip(self.cache, next(self.child_iterator)):
```

Comment:
This `child_iterator` should at least be documented as an attribute. It is unclear from the `CachedDataStream` code what it is and when it was created.